File: packages/guards/src/rate-limit.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

interface RateLimitOptions {
  action: 'block' | 'warn';
  maxRequests: number;
  windowMs: number;
  keyFn?: (text: string, ctx: GuardContext) => string;
}

export function rateLimit(options: RateLimitOptions): Guard {
  const buckets = new Map<string, number[]>();
  const keyFn = options.keyFn ?? (() => 'global');

  function prune(key: string, now: number): number[] {
    const timestamps = buckets.get(key) ?? [];
    const cutoff = now - options.windowMs;
    const valid = timestamps.filter((t) => t > cutoff);
    buckets.set(key, valid);
    return valid;
  }

  return {
    name: 'rate-limit',
    version: '0.1.0',
    description: 'In-memory sliding window rate limiter',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(
      text: string,
      ctx: GuardContext,
    ): Promise<GuardResult> {
      const start = performance.now();
      const key = keyFn(text, ctx);
      const now = Date.now();
      const window = prune(key, now);
      const triggered = window.length >= options.maxRequests;

      if (!triggered) {
        window.push(now);
      }

      return {
        guardName: 'rate-limit',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: {
          key,
          requestsInWindow: window.length,
          maxRequests: options.maxRequests,
        },
      };
    },
  };
}

File: python/open_guardrail/guards/rate_limit.py
"""Simple in-memory rate limiter."""

import time
from typing import List

from open_guardrail.core import GuardResult


class _RateLimit:
    def __init__(
        self, *, action: str = "block", max_requests: int = 100, window_ms: int = 60000,
    ) -> None:
        self.name = "rate-limit"
        self.action = action
        self.max_requests = max_requests
        self.window_ms = window_ms
        self._timestamps: List[float] = []

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        now = time.time() * 1000
        cutoff = now - self.window_ms
        self._timestamps = [t for t in self._timestamps if t > cutoff]
        self._timestamps.append(now)
        triggered = len(self._timestamps) > self.max_requests
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name=self.name,
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"Rate limit exceeded: {len(self._timestamps)}/{self.max_requests} in {self.window_ms}ms" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"current_count": len(self._timestamps), "max_requests": self.max_requests} if triggered else None,
        )


def rate_limit(
    *, action: str = "block", max_requests: int = 100, window_ms: int = 60000,
) -> _RateLimit:
    return _RateLimit(action=action, max_requests=max_requests, window_ms=window_ms)
